atomic_reactor/koji_util.py

```diff
@@ -271,3 +284,336 @@ def get_koji_module_build(session, module_spec):
+        return Output(file=logfile, metadata=metadata)
+
+    def add_log_type(output, arch):
+        logfile, metadata = output
+        metadata.update({'type': 'log', 'arch': arch})
+        return Output(file=logfile, metadata=metadata)
+
+    extra_output_file = None
+    output_files = []
+
+    arch = os.uname()[4]
+
+    if source_build:
+        workflow.builder.tasker.pull_image(pullspec)
+        defer_removal(workflow, pullspec)
+        image_inspect = workflow.builder.tasker.inspect_image(pullspec)
+        image_id = image_inspect['Id']
+
+        history = workflow.builder.tasker.get_image_history(image_id)
+        diff_ids = image_inspect[INSPECT_ROOTFS].get(INSPECT_ROOTFS_LAYERS, ())
+
+        # diff_ids is ordered oldest first
+        # history is ordered newest first
+        # We want layer_sizes to be ordered oldest first
+        layer_sizes = [{"diff_id": diff_id, "size": layer['Size']}
+                       for (diff_id, layer) in zip(diff_ids, reversed(history))]
+
+        platform = arch
+
+    else:
+        output_files = [add_log_type(add_buildroot_id(metadata), arch)
+                        for metadata in logs]
+
+        # Parent of squashed built image is base image
+        image_id = workflow.builder.image_id
+        parent_id = None
+        if not workflow.builder.base_from_scratch:
+            parent_id = workflow.builder.base_image_inspect['Id']
+
+        layer_sizes = workflow.layer_sizes
+
+    registries = workflow.push_conf.docker_registries
+    config = copy.deepcopy(registries[0].config)
+
+    # We don't need container_config section
+    if config and 'container_config' in config:
+        del config['container_config']
+
+    repositories, typed_digests = get_repositories_and_digests(workflow, pullspec)
+
+    tags = set(image.tag for image in workflow.tag_conf.images)
+    metadata, output = get_image_output(workflow, image_id, platform)
+
+    metadata.update({
+        'arch': arch,
+        'type': 'docker-image',
+        'components': [],
+        'extra': {
+            'image': {
+                'arch': arch,
+            },
+            'docker': {
+                'id': image_id,
+                'repositories': repositories,
+                'layer_sizes': layer_sizes,
+                'tags': list(tags),
+                'config': config,
+                'digests': typed_digests,
+            },
+        },
+    })
+
+    if not config:
+        del metadata['extra']['docker']['config']
+
+    if not source_build:
+        metadata['components'] = get_image_components(workflow)
+
+        if not workflow.builder.base_from_scratch:
+            metadata['extra']['docker']['parent_id'] = parent_id
+
+    # Add the 'docker save' image to the output
+    image = add_buildroot_id(output)
+    if source_build:
+        output_files.append(metadata)
+        extra_output_file = output
+    else:
+        output_files.append(image)
+
```

Comment:
Could you please explain this section to me? Why it differs for source containers?

Reply:
for build container
worker already uploaded image, so orchestrator needs only metadata

for source container
we have just "orchestrator" so we need metadata + also image to upload

Reply:
can we have comment there?

Reply:
I'm also unhappy, that the helper module decides where to fetch data based on build type. But it's not an issue of this PR

Reply:
sure